ReSharper.FSharp/src/FSharp/FSharp.Psi.Intentions/src/Intentions/FunctionAnnotationAction.fs

```diff
@@ -221,36 +231,61 @@ module SpecifyTypes =
-            for fcsParamGroup, parameter in Seq.zip fcsParamGroups parameters do
+        let rec specifyParameterTypes (fcsParamGroups: 'a seq) (getFcsType: 'a -> FSharpType) (enumerate: 'a -> 'a seq) (parameters: IFSharpPattern seq) acc isTopLevel =
+            Seq.zip fcsParamGroups parameters
+            |> Seq.fold (fun acc (fcsParamGroup, parameter) ->
                 match parameter.IgnoreInnerParens() with
-                | :? IConstPat | :? ITypedPat -> ()
+                | :? IConstPat | :? ITypedPat -> acc
                 | TupleLikePattern pat when isTopLevel ->
                     let fcsTypes = enumerate fcsParamGroup
-                    specifyParameterTypes fcsTypes getFcsType enumerate pat.Patterns false
+                    specifyParameterTypes fcsTypes getFcsType enumerate pat.Patterns acc false
                 | pattern ->
                     let fcsType = getFcsType fcsParamGroup
-                    specifyPatternType displayContext fcsType pattern
+                    specifyPatternTypeImpl displayContext fcsType pattern :: acc
+            ) acc
 
         let parameters = decl.ParametersDeclarations |> Seq.map _.Pattern
 
         if mfv.IsMember then
             let enumerate x = x |> Seq.map (fun x -> [|x|] :> IList<_>)
             let types = mfv.CurriedParameterGroups
-            specifyParameterTypes types (fun x -> x[0].Type) enumerate parameters true
+            specifyParameterTypes types (fun x -> x[0].Type) enumerate parameters [] true
         else
             let types = getFunctionTypeArgs true mfv.FullType
-            specifyParameterTypes types id (_.GenericArguments) parameters true
-
-    let rec specifyTypes (node: ITreeNode) (availability: Availability) =
+            specifyParameterTypes types id (_.GenericArguments) parameters [] true
+
+    let private postProcessAnnotations (file: IFSharpFile) annotationsInfo =
+        let annotationsInfo =
+            annotationsInfo
+            |> Seq.fold collectTypeUsages []
+            |> Seq.map (fun (typeUsage, fcsType, context: ITreeNode) ->
+                let typeReference = typeUsage.ReferenceName
+                let reference = typeReference.Reference.AllowAllSymbolCandidatesCheck()
+                let fcsSymbol = fcsType.TypeDefinition
+                let declaredElement = fcsSymbol.GetDeclaredElement(context.GetPsiModule()).As<IClrDeclaredElement>()
+                context, reference, fcsSymbol, declaredElement)
+            |> Seq.filter (fun (_, reference, _, declaredElement) -> isNotNull reference && isNotNull declaredElement)
+            |> Seq.toArray
+
+        for context, reference, _, declaredElement in annotationsInfo do
+            reference.SetRequiredQualifiers(declaredElement, context)
+
+        use pinCheckResultsCookie = file.PinTypeCheckResults(true, "Specify types")
```

Comment:
Would it work if the results are pinned before the modifications?

Reply:
And maybe we should move this out of this PR and try to do this more systematically instead.

Reply:
It's fine to have this as is, without the assertions, if we want to add them next. I just want it to make it easier to revert or cherry-pick this PR if needed. Working on such assertions sounds like potentially unrelated changes here and there that would be nice to avoid.